the-collab-lab/tcl-75-smart-shopping-list
# [ISSUE 13] -  view shopping list items in order of how soon they to need to be bought again

## Description

Attempt to encapsulate calculating urgency with which an item should be bought. These changes ensure more efficient state management, removing global urgencyObject in favor of abstracted state within useUrgency hook. Display items pre-sorted by dateLastPurchased(overdue), then dateNextPurchased(urgency), then alphabetically(if similar urgency).

**What and Why**

The hook is initializing an initial state with five sets that will store items sorted by urgency(overdue, soon, kind of soon, not soon, inactive). Items that are accepted as props go through a logical check that defines when they were last purchased and whether they are inactive, then when they should be purchased next and how they should get sorted further. After all items are sorted, all sets are converted to alphabetically sorted arrays. With this object we set a state that is passed to ListItem component as its urgencyStatus. Based on the status a relevant UI is set.
I chose this approach as I couldn't figure out how to sort alphabetically after the filtering algorithm. I had to come up with a different solution that involved centralized state management and alphabetical pre-sorting. I opted to sort all items by urgency, then sorting by alphabet and then display a flattened array of all items.

**Open questions:**

- I noticed that if an item is clicked on very many times, the code keeps on indicating that it is 'overdue', because it was purchased today and needs to be purchased today again. This makes me wonder whether the code does not work as expected, or whether a user should have an ability to reset dateNextPurchased.
- Would be happy to hear how to better style urgency UI.

<table border="1">
  <thead>
    <tr>
      <th>Urgency Status</th>
      <th>MUI Title</th>
      <th width="100">Icon</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>overdue</td>
      <td>Restore</td>
      <td>
 <img src="https://github.com/user-attachments/assets/c7cadd0e-dbf8-4f58-b304-a929e982c447" alt="Restore" />
</td>
    </tr>
    <tr>
      <td>soon</td>
      <td>RestartAlt</td>
      <td>
 <img src="https://github.com/user-attachments/assets/a0943a55-72f0-47e2-bb35-916ff01d74ae" alt="RestartAlt" />
</td>
    </tr>
    <tr>
      <td>kindOfSoon</td>
      <td>RadioButtonUnchecked</td>
      <td>
 <img src="https://github.com/user-attachments/assets/3d98ba44-a06d-4cb1-bd66-0b0dd7901047" alt="RadioButtonUnchecked" />
</td>
    </tr>
    <tr>
      <td>notSoon</td>
      <td>RemoveCircle</td>
      <td>
 <img src="https://github.com/user-attachments/assets/429a96c0-cf76-4c1b-85c7-939ff59099c7" alt="RemoveCircle" />
</td>
    </tr>
    <tr>
      <td>inactive</td>
      <td>RadioButtonChecked</td>
      <td>
 <img src="https://github.com/user-attachments/assets/87e920f3-de85-437f-b406-4f69894b99cb" alt="RadioButtonChecked" />
</td>
    </tr>
  </tbody>
</table>



## Related Issue

#13 

## Acceptance Criteria

- [x]  the code lives in the hooks.js as it is a hook.

## Type of Changes

`enhancement`

## Updates

### Before

![Screenshot 2024-10-03 092233](https://github.com/user-attachments/assets/7eb5c3c0-6120-4908-937b-80c809b78006)

### After

![Screenshot 2024-10-03 191751](https://github.com/user-attachments/assets/0384a4bf-5239-4ec3-b639-b7d893b896dd)

![sort-by-urgency](https://github.com/user-attachments/assets/a493dfc3-ce38-48fa-9009-bfee37336795)

## Testing Steps / QA Criteria

1. run npm start and sign in
2. add new items
3. click on purchase checkbox to ensure that items get sorted by urgency
